DPL: refactor ConfigParamRegistry to use C++20

ktf opened the pull request:
Use constraints to make sure we can out-of-line known
types of parameter.

Remove the need to include Array2D.h when not used.

Reduce dependence on ptree, in particular moving the helper methods
out of line.

alibuild commented:
Error while checking build/O2/fullCI for 48757bd6de908f5af7441d60ff2b80c28dff4bb8 at 2025-01-10 13:36:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13850/48757bd6de908f5af7441d60ff2b80c28dff4bb8/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI for c4a029dd6c665c12daeb5ad927d2462fd6e2e983 at 2025-01-10 14:58:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13850-slc8_x86-64/0/Framework/Core/include/Framework/ConfigParamRegistry.h:96:44: error: cannot convert 'boost::property_tree::basic_ptree<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >' to 'long long int' in initialization
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13850/c4a029dd6c665c12daeb5ad927d2462fd6e2e983/build_O2_fullCI/pretty.html).

ktf committed:
DPL: refactor ConfigParamRegistry to use C++20

Use constraints to make sure we can out-of-line known
types of parameter.

Remove the need to include Array2D.h when not used.

Reduce dependence on ptree, in particular moving the helper methods
out of line.

ktf commented:
macOS tested separately. Merging.
